internal/uvm/scsi.go

```diff
@@ -60,38 +63,78 @@ func (uvm *UtilityVM) findSCSIAttachment(findThisHostPath string) (int, int32, s
 }
 
 // AddSCSI adds a SCSI disk to a utility VM at the next available location.
+// This function should be called for a RW/scratch layer. For read-only layers
+// on LCOW as an alternate to PMEM for large layers, use AddSCSILayer instead.
+//
+// hostPath is required
+// uvmPath is optional.
+func (uvm *UtilityVM) AddSCSI(hostPath string, uvmPath string) (int, int32, error) {
+	return uvm.addSCSIActual(hostPath, uvmPath, false, false)
+}
+
+// AddSCSILayer adds a read-only layer disk to a utility VM at the next available
+// location. This function is used by LCOW as an alternate to PMEM for large layers.
+// The UVMPath will always be /tmp/S<controller>/<lun>.
+func (uvm *UtilityVM) AddSCSILayer(hostPath string) (int, int32, error) {
+	return uvm.addSCSIActual(hostPath, "", true, true)
+}
+
+// addSCSIActual is the implementation behind the external functions AddSCSI and
+// AddSCSILayer.
 //
 // We are in control of everything ourselves. Hence we have ref-
 // counting and so-on tracking what SCSI locations are available or used.
 //
 // hostPath is required
 // uvmPath is optional.
+// isLayer indicates that this is a read-only (LCOW) layer VHD too big to fit on vpmem.
+// calculateUVMPath indicates that the UVM path is calculated based on /tmp/SC/L
 //
 // Returns the controller ID (0..3) and LUN (0..63) where the disk is attached.
-func (uvm *UtilityVM) AddSCSI(hostPath string, uvmPath string) (int, int32, error) {
+func (uvm *UtilityVM) addSCSIActual(hostPath string, uvmPath string, isLayer bool, calculateUVMPath bool) (int, int32, error) {
 	if uvm == nil {
 		return -1, -1, ErrNoUvmParameter
 	}
+	if uvm.operatingSystem == "windows" && isLayer {
+		return -1, -1, ErrSCSILayerWCOWUnsupported
+	}
 	logrus.Debugf("uvm::AddSCSI id:%s hostPath:%s uvmPath:%s", uvm.id, hostPath, uvmPath)
 
 	if uvm.scsiControllerCount == 0 {
 		return -1, -1, ErrNoSCSIControllers
 	}
 
 	uvm.m.Lock()
-	if _, _, _, err := uvm.findSCSIAttachment(hostPath); err == nil {
-		uvm.m.Unlock()
-		return -1, -1, ErrAlreadyAttached
+	if controller, lun, _, err := uvm.findSCSIAttachment(hostPath); err == nil {
+		// So is attached
+		if isLayer {
+			// Increment the refcount
+			uvm.scsiLocations[controller][lun].refCount++
+			logrus.Debugf("uvm::AddSCSI id:%s hostPath:%s uvmPath:%s refCount now %d", uvm.id, hostPath, uvmPath, uvm.scsiLocations[controller][lun].refCount)
+			uvm.m.Unlock()
+			return controller, int32(lun), nil
+		} else {
+			uvm.m.Unlock()
+			return -1, -1, ErrAlreadyAttached
+		}
 	}
-	uvm.m.Unlock()
 
-	controller, lun, err := uvm.allocateSCSI(hostPath, uvmPath)
+	// At this point, we know it's not attached, regardless of whether it's ref-counted layer VHD, or not.
+	controller, lun, err := uvm.allocateSCSI(hostPath, uvmPath, isLayer)
 	if err != nil {
+		uvm.m.Unlock()
 		return -1, -1, err
 	}
 
+	if calculateUVMPath && uvmPath == "" {
```

Comment:
Just move this into `AllocateSCSILayer`

Reply:
Ah, that's not so easy without duplicating a bunch of logic. I won't know the path until it's been determined whether it's already attached, or if not already attached, after allocating an address. All that logic is currently in this function and common to both paths.

Reply:
That said, I can probably remove `calculateUVMPath` and assume it's calculated for layers. Would that work for you?

Reply:
Have updated it to remove `calculateUVMPath`